File: c++/count_vowel_substrings_of_a_string.cpp
#include <cassert>
#include <iostream>
#include <string>
#include <unordered_set>
using namespace std;

class Solution {
   public:
    int countVowelSubstrings(std::string word) {
        std::unordered_set<char> vowels = {'a', 'e', 'i', 'o', 'u'};
        int result = 0;
        std::unordered_map<char, int> vowelIds;
        int start = 0;
        int n = word.length();

        for (int i = 0; i < n; i++) {
            char c = word[i];

            if (vowels.find(c) == vowels.end()) {
                vowelIds.clear();
                continue;
            }

            if (vowelIds.empty()) start = i;

            vowelIds[c] = i;

            if (vowelIds.size() == 5) {
                int minIdx = n;
                for (const auto& kv : vowelIds) minIdx = std::min(minIdx, kv.second);
                result += minIdx - start + 1;
            }
        }

        return result;
    }
};

int main() {
    Solution sol;
    assert(sol.countVowelSubstrings("aeiouu") == 2);
    assert(sol.countVowelSubstrings("unicornarihan") == 0);
    assert(sol.countVowelSubstrings("cuaieuouac") == 7);
    cout << "All test cases passed!" << endl;
    return 0;
}

File: python/count_vowel_substrings_of_a_string.py
import collections
from typing import Set


class Solution:
    def countVowelSubstrings(self, word: str) -> int:
        vowels = ("a", "e", "i", "o", "u")

        result = 0
        start = 0
        vowel_idx = {}
        for idx, c in enumerate(word):
            if c in vowels:
                if not vowel_idx:
                    start = idx
                vowel_idx[c] = idx
                if len(vowel_idx) == 5:
                    result += min(vowel_idx.values()) - start + 1
            elif vowel_idx:
                vowel_idx = {}

        return result


sol = Solution()
word = "aeiouu"
print(sol.countVowelSubstrings(word), 2)
word = "unicornarihan"
print(sol.countVowelSubstrings(word), 0)
word = "cuaieuouac"
print(sol.countVowelSubstrings(word), 7)
word = "bbaeixoubb"
print(sol.countVowelSubstrings(word), 0)
